sdk/core/Azure.Core/src/Shared/ClientDiagnostics.cs

```diff
@@ -2,52 +2,154 @@
+            ValueTask<string> messageTask = CreateRequestFailedMessageAsync(response, message, errorCode, additionalInfo, false);
+            Debug.Assert(messageTask.IsCompleted);
+            return messageTask.GetAwaiter().GetResult();
+        }
+
+        private async ValueTask<string> CreateRequestFailedMessageAsync(Response response, string? message, string? errorCode, IDictionary<string, string>? additionalInfo, bool async)
+        {
+            var content = await ReadContentAsync(response, async).ConfigureAwait(false);
+
+            return CreateRequestFailedMessageWithContent(response, message, content, errorCode, additionalInfo);
+        }
+        public string CreateRequestFailedMessageWithContent(Response response, string? message, string? content, string? errorCode, IDictionary<string, string>? additionalInfo)
+        {
+            StringBuilder messageBuilder = new StringBuilder()
+                .AppendLine(message ?? DefaultMessage)
+                .Append("Status: ")
+                .Append(response.Status.ToString(CultureInfo.InvariantCulture))
+                .Append(" (")
+                .Append(response.ReasonPhrase)
+                .AppendLine(")");
+
+            if (!string.IsNullOrWhiteSpace(errorCode))
+            {
+                messageBuilder.Append("ErrorCode: ")
+                    .Append(errorCode)
+                    .AppendLine();
+            }
+
+            if (additionalInfo != null && additionalInfo.Count > 0)
+            {
+                messageBuilder
+                    .AppendLine()
+                    .AppendLine("Additional Information:");
+                foreach (KeyValuePair<string, string> info in additionalInfo)
+                {
+                    messageBuilder
+                        .Append(info.Key)
+                        .Append(": ")
+                        .AppendLine(info.Value);
+                }
+            }
+
+            if (content != null)
             {
-                return default;
+                messageBuilder
+                    .AppendLine()
+                    .AppendLine("Content:")
+                    .AppendLine(content);
             }
-            var scope = new DiagnosticScope(name, _source);
 
-            if (_resourceProviderNamespace != null)
+            messageBuilder
+                .AppendLine()
+                .AppendLine("Headers:");
+
+            foreach (HttpHeader responseHeader in response.Headers)
             {
-                scope.AddAttribute("az.namespace", _resourceProviderNamespace);
+                string headerValue = _sanitizer.SanitizeHeader(responseHeader.Name, responseHeader.Value);
+                messageBuilder.AppendLine($"{responseHeader.Name}: {headerValue}");
             }
-            return scope;
+
+            return messageBuilder.ToString();
+        }
+
+        private static async ValueTask<string?> ReadContentAsync(Response response, bool async)
```

Comment:
Are we using `async` or `isAsync` as a name for the variable?

Reply:
Also, do we need to pass `CancellationToken` here?

Reply:
> Are we using async or isAsync as a name for the variable?

I don't think we ever settled on one name for this parameter but that would be nice.

> Also, do we need to pass CancellationToken here?

Yep, filed https://github.com/Azure/azure-sdk-for-net/issues/9792